rfcs/text/0005_route_handler.md

```diff
@@ -0,0 +1,184 @@
+Rather `Route Handler` in New platform has the next signature:
+```typescript
+type RequestHandler = (
+  context: Context,
+  request: KibanaRequest,
+  t: KibanaResponseToolkit
+) => KibanaResponse | Promise<KibanaResponse>;
+```
+and accepts next Kibana specific parameters as arguments:
+- context: [Context](https://github.com/elastic/kibana/blob/master/rfcs/text/0003_handler_interface.md). A handler context contains core service and plugin functionality already scoped to the incoming request.
+- request: [KibanaRequest](https://github.com/elastic/kibana/blob/master/src/core/server/http/router/request.ts). An immutable representation of the incoming request details, such as body, parameters, query, url and route information. Note: you **must** to specify route schema during route declaration to have access to `body, parameters, query` in the request object. You  cannot extend KibanaRequest with arbitrary data nor remove any properties from it.
+```typescript
+interface KibanaRequest {
+  url: url.Url;
+  headers: Record<string, string | string [] | undefined>;
+  params?: Record<string, any>;
+  body?: Record<string, any>;
+  query?: Record<string, any>;
+  route: {
+    path: string;
+    method: 'get' | 'post' | ...
+    options: {
+      authRequired: boolean;
+      tags: string [];
+    }
+  }
+}
+```
+-  t: [KibanaResponseToolkit](https://github.com/elastic/kibana/blob/master/src/core/server/http/router/response.ts#L27)
+Provides a set of pre-configured methods to respond to an incoming request. It is expected that handler **always** returns a result of one of `KibanaResponseToolkit` methods as an output:
+```typescript
+interface KibanaResponseToolkit {
+  [method:string]: (...params: any) => KibanaResponse
+}
+router.get(...,
+  (context: Context, request: KibanaRequest, t: KibanaResponseToolkit): KibanaResponse => {
+   return t.ok();
+   // or
+   return t.redirected('/url');
+   // or
+   return t.badRequest(error);
+);
+```
+*KibanaResponseToolkit*  methods allow an end user to adjust the next response parameters:
+- Body. Supported values:`undefined | string | JSONValue | Buffer | Stream`.
+- Status code. 
+- Headers. Supports adjusting [known values](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v10/http.d.ts#L8) and attaching [custom values as well](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v10/http.d.ts#L67)
+
+Other response parameters, such as `etag`, `MIME-type`, `bytes` that used in the Legacy platform could be adjusted via Headers.
+
+The router handler doesn't expect that logic inside can throw or return something different from `KibanaResponse`. In this case, Http service will respond with `Server error` to prevent exposure of internal logic details.
```

Comment:
If a handler throws a `KibanaResponse`, the server should still use this thrown response to populate the actual HTTP response. Any other types of objects thrown should result in a 500.

It's not clear whether or not this is the case in this design.

Reply:
A handler function should:
- do not throw an error.
- always return a `KibanaResponse`.

Reply:
Interesting, so with this approach you are encouraged to construct the response in the handler code itself, rather than allow business logic to construct HTTP responses.

This encourages less coupling of our logic to the HTTP framework, which is a big plus.

```ts
// Without allowing exceptions (this RFC)
const async myHandler(context, req, t) {
  try {
    const obj = await retrieveObj(context.savedObjectClient, req.params.id);
    const res = processObj(obj);
    return t.ok(res);
  } catch (e) {
    // handler constructs the error response
    return t.notFound();
  }
}
```

```ts
// Allowing exceptions (not this RFC - more like Hapi)
const async myHandler(context, req, t) {
  // business logic construct and throw a 404 on it's own
  const obj = await retrieveObj(context.savedObjectClient, req.params.id, t);
  const res = processObj(obj);
  return t.ok(res);
}
```

Reply:
- Throwing and re-throwing errors inside the code is not fun and hard to follow.
- If we mix all exceptions it's hard to type properly and distinguish between business errors & exception raised due to a bug.
- Core doesn't dictate how to handle errors anymore. If plugin wants to use `Boom` - ok, but it's not a part of core contract anymore.

I'd write your example as:
```js
const async myHandler(context, req, t) {
    const obj = await retrieveObj(context.savedObjectClient, req.params.id);
    if (!obj) return t.notFound();
    const res = processObj(obj);
    return t.ok(res);
}
```

Reply:
I like that exceptions are used for exceptional or unexpected behaviour so I prefer the API of the RFC above the way Hapi does it :+1:

Reply:
> Throwing and re-throwing errors inside the code are not fun and hard to follow.

++ I like how this removes a small element of "magic" and forces you to explicitly respond with errors in the handler itself. To me the approach in this RFC feels simpler, even if it isn't the Hapi way of doing things.

Reply:
Sounds like we're all on the same page here 🎉 . I just wanted to be clear on how this would work in the NP.